src/transport/quinn.rs

```diff
@@ -23,17 +23,24 @@ type Socket<In, Out> = (SendSink<Out>, RecvStream<In>);
 const MAX_FRAME_LENGTH: usize = 1024 * 1024 * 16;
 
 #[derive(Debug)]
-struct ServerChannelInner {
+struct ServerEndpointInner {
     endpoint: Option<quinn::Endpoint>,
     task: Option<tokio::task::JoinHandle<()>>,
     local_addr: [LocalAddr; 1],
     receiver: flume::Receiver<SocketInner>,
 }
 
-impl Drop for ServerChannelInner {
+impl Drop for ServerEndpointInner {
     fn drop(&mut self) {
+        tracing::debug!("Dropping server endpoint");
         if let Some(endpoint) = self.endpoint.take() {
-            endpoint.close(0u32.into(), b"server channel dropped");
+            tracing::debug!("Closing endpoint");
+            endpoint.close(0u32.into(), b"server endpoint dropped");
+            // spawn a task to wait for the endpoint to notify peers that it is closing
+            tokio::spawn(async move {
+                endpoint.wait_idle().await;
+                tracing::debug!("Endpoint closed");
+            });
```

Comment:
how come i've never thought of spawning tasks in drop before?  nice hack

Reply:
It's horrible. But what can you do if you live in async land - we already learned how to clean up misbehaving tokio background tasks...

Reply:
I both enjoy and am horrified of this..